numpy/_core/src/multiarray/datetime.c

```diff
@@ -2059,18 +2059,10 @@ NpyDatetime_ConvertPyDateTimeToDatetimeStruct(
     out->month = 1;
     out->day = 1;
 
-    /* Need at least year/month/day attributes */
-    if (!PyObject_HasAttrString(obj, "year") ||
-            !PyObject_HasAttrString(obj, "month") ||
-            !PyObject_HasAttrString(obj, "day")) {
-        return 1;
-    }
-
     /* Get the year */
     /* Checking if Exception will be raised it would be -1 else + value */
-    int res = PyObject_GetOptionalAttr(obj, npy_interned_str.year, &tmp);
-    if (res != 1) {
-        return -1;
+    if (PyObject_GetOptionalAttr(obj, npy_interned_str.year, &tmp) != 1) {
+        return 1;
```

Comment:
If this returns -1, you need to return -1.

Reply:
Hi @eric-wieser , sorry If I may be getting it wrong. 

But the check which I removed 

`if (!PyObject_HasAttrString(obj, "year") ||
            !PyObject_HasAttrString(obj, "month") ||
            !PyObject_HasAttrString(obj, "day")) {
        return 1;
    }`

was returning 1 if any of this attribute is not present in the obj, now with PyObject_GetOptionalAttr if this attribute is not found this method will negative value, and method presume to return 1 for these non happy cases. 

If I will return -1, it will fails few of the test cases ( 4 as of now test_object_casts_NULL_None_equivalence, :test_datetime_scalar_construction, :test_pyobject_roundtrip, test_corecursive_input ). 

Please help me understand, where I am getting this wrong ?

Reply:
i think, i got the direction you were referring to.